### ISSUE TITLE
Rename regexManagers to customManagers

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#19066]
### Describe the proposed change(s).

- rename `regexManagers` to `customManagers`
- add a new field `customType` which defaults to `regex`
- add migration code

**Comment by rarkins** [comment]
Nest the existing regexManagers code under a `regex/` subdirectory too and wrap them so that we're ready for new customTypes

**Comment by rarkins** [comment]
1. Yes
2. Wrap/duplicate the functions a little

**Comment by viceice** [comment]
use 2.

**Comment by reitzig** [comment]
Will there be any API and/or implementation changes? (Sorry if the answer is obvious, didn't check the PR.)

FWIW, I was prompted to share thoughts on how regexManagers could be made easier to debug: [renovatebot/renovate#20881](https://github.com/renovatebot/renovate/discussions/20881#discussioncomment-5595438)

**Comment by RahulGautamSingh** [comment]
We will need to migrate `matchManagers` as well. How should it be migrated `regex` to `custom` or let it be `regex`
I prefer `regex` because if user choosed `custom` then incase we have multiple custom managers all will be matched.

**Comment by rarkins** [comment]
Support both then. custom will match any manager while regex matches only Regex, jsonata only jsonata, etc

**Comment by RahulGautamSingh** [comment]
Doables:
- [ ] move regex manager in `custom/regex` https://github.com/renovatebot/renovate/pull/23715
- [ ] add custom manager api https://github.com/renovatebot/renovate/pull/23357
- [ ] implement `custom.*` syntax for `enabledManagers` and `matchManagers`
- [ ] add `customType` field to regexManagers config
- [ ] rename `regexManagers` to `customManagers`

**Comment by RahulGautamSingh** [comment]
After discussions in [here](https://github.com/renovatebot/renovate/pull/23357#discussion_r1266334503), we support the `custom*` syntax, 
Example:
```ts
matchManagers = ['custom.regex', 'custom.jsonata']  // allow custom managers separately
enabledManagers= ['custom.*'] // allows all custom managers
```

**Comment by RahulGautamSingh** [comment]
Doables in order:
- [ ] move regex manager in `custom/regex` https://github.com/renovatebot/renovate/pull/23715
- [ ] add custom manager api https://github.com/renovatebot/renovate/pull/23357
- [ ] implement `custom.*` syntax for `enabledManagers` and `matchManagers`
- [ ] add `customType` field to regexManagers config
- [ ] rename `regexManagers` to `customManagers`

**Comment by RahulGautamSingh** [comment]
> Support both then. custom will match any manager while regex matches only Regex, jsonata only jsonata, etc

After discussions in https://github.com/renovatebot/renovate/pull/23357#discussion_r1266334503, we have decided to support the `custom*` syntax, 
Example:
```ts
matchManagers = ['custom.regex', 'custom.jsonata']  // custom managers separately
enabledManagers= ['custom.*'] // all custom managers
```

**Comment by RahulGautamSingh** [comment]
Doables in order:
- [ ] move regex manager in `custom/regex` https://github.com/renovatebot/renovate/pull/23715
- [ ] add custom manager api https://github.com/renovatebot/renovate/pull/23357
- [ ] implement `custom.*` syntax for `enabledManagers` and `matchManagers`
- [ ] add `customType` field to regexManagers config
- [ ] rename `regexManagers` to `customManagers`

**Comment by RahulGautamSingh** [comment]
We will need separate interfaces for `CustomManager` , `RegexManager` ...
`CustomManager` : will contain the common fields that  are expected in all custom managers eg. `customType`, `fileMatch`, and `RegexManager` : will contain the fields which are unique to it like `matchStrings` , `matchStringsStrategy` & its templates

Similarly we will need to handle validation based on `customType` as well , after validating the `customType` and `fileMatch` fields.
For eg. 
```ts
 if (key === 'regexManagers') {
 ...
                } else if (is.nonEmptyString(regexManager.customType)) {
                  if (is.nonEmptyArray(regexManager.fileMatch)) {
                    // TODO: Add validation as per customType, preferrably a separate file or function for readability
                    switch (regexManager.customType) {
                      case 'regex':
                        errors.push(
                          ...(validateRegexManagerFields(
                            regexManager,
                            currentPath
                          ) as any)
                        );
                        break;
                    }
                  } else {
                    errors.push({
                      topic: 'Configuration Error',
                      message: `Each Regex Manager must contain a non-empty fileMatch array`,
                    });
                  }
                } else {
                  errors.push({
                    topic: 'Configuration Error',
                    message: `Each Regex Manager must contain a non-empty customType string`,
                  });
                }
              }
            }
... 
```

Hence, when a new custom manager is added, we will need a new interface and a new validation function to validate its fields separately.

**Comment by RahulGautamSingh** [comment]
Should we rename `RegexManagerPresets` to `CustomManagerPresets` ? It makes sense, since in future we might include presets of other custom managers to the same list.

**Comment by RahulGautamSingh** [comment]
We can close this issue now, and transfer the enhancements ie. implementing `custom.*` syntax for `enabledManagers` and `matchManagers` to https://github.com/renovatebot/renovate/issues/21760

**Comment by RahulGautamSingh** [comment]
> Doables in order:
> 
> * [x]  move regex manager in `custom/regex` [refactor(managers): move regex in `custom` folder #23715](https://github.com/renovatebot/renovate/pull/23715)
> * [x]  add custom manager api [refactor(managers): implement custom managers api #23357](https://github.com/renovatebot/renovate/pull/23357)
> * [x]  add `customType` field to regexManagers config [feat(managers/custom): introduce `customType` field #23919](https://github.com/renovatebot/renovate/pull/23919)
> * [x]  rename `regexManagers` to `customManagers` [feat(migrate): `regexManagers` -> `customManagers` #24451](https://github.com/renovatebot/renovate/pull/24451)
> 
> Enhancement not directly related to the implementation
> 
> * [ ]  implement `custom.*` syntax for `enabledManagers` and `matchManagers`
>   1. [feat: implement `custom.<customMgrName>` syntax for `matchManagers` #24112](https://github.com/renovatebot/renovate/pull/24112)
>   2. [feat: `custom.<customMgrName>` syntax for enabledManagers #24079](https://github.com/renovatebot/renovate/pull/24079)

We can close this issue now, and transfer the enhancements ie. implementing `custom.*` syntax for `enabledManagers` and `matchManagers` to https://github.com/renovatebot/renovate/issues/21760

**Comment by TWiStErRob** [comment]
Note: this is not mentioned in https://docs.renovatebot.com/release-notes-for-major-versions/ for v37

**Comment by rarkins** [comment]
@TWiStErRob any thing with migrations is not treated as a breaking change

**Comment by TWiStErRob** [comment]
@rarkins I see it's not breaking, but could still worth a mention in commentary as it's a concept change?

**Comment by rarkins** [comment]
We document changes in whatever release they fall in

**Comment by TWiStErRob** [comment]
The last PR that was the actual rename (#24451) was landed in 36.107.0, does that not make it part of major v37?

**Comment by rarkins** [comment]
Yes, just like the 106 features released before it in v36. We can't mention them all in v37
